drivers/bus/isapnp/hardware.c

```diff
@@ -306,320 +430,1311 @@ IsaPnpChecksum(
+
+        if (!LogDevice->MemRange[i].CurrentBase)
+            break;
+
+        LogDevice->MemRange[i].CurrentLength = ReadMemoryLimit(ReadDataPort, i) << 8;
+
+        if (ReadMemoryControl(ReadDataPort, i) & MEMORY_UPPER_LIMIT)
+        {
+            LogDevice->MemRange[i].CurrentLength -= LogDevice->MemRange[i].CurrentBase;
+        }
+        else
+        {
+            LogDevice->MemRange[i].CurrentLength =
+                RANGE_LENGTH_TO_LENGTH(LogDevice->MemRange[i].CurrentLength);
+        }
+    }
+    for (i = 0; i < RTL_NUMBER_OF(LogDevice->MemRange32); i++)
+    {
+        LogDevice->MemRange32[i].CurrentBase = ReadMemoryBase32(ReadDataPort, i);
+
+        if (!LogDevice->MemRange32[i].CurrentBase)
+            break;
+
+        LogDevice->MemRange32[i].CurrentLength = ReadMemoryLimit32(ReadDataPort, i);
+
+        if (ReadMemoryControl32(ReadDataPort, i) & MEMORY_UPPER_LIMIT)
+        {
+            LogDevice->MemRange32[i].CurrentLength -= LogDevice->MemRange32[i].CurrentBase;
+        }
+        else
+        {
+            LogDevice->MemRange32[i].CurrentLength =
+                RANGE_LENGTH_TO_LENGTH(LogDevice->MemRange32[i].CurrentLength);
+        }
+    }
+
+    LogDevice->Flags |= ISAPNP_HAS_RESOURCES;
+    return TRUE;
+}
+
+static
+CODE_SEG("PAGE")
+VOID
+ResetCurrentResources(
+    _In_ PISAPNP_LOGICAL_DEVICE LogDevice)
+{
+    UCHAR i;
+
+    PAGED_CODE();
+
+    LogDevice->Flags &= ~ISAPNP_HAS_RESOURCES;
+
+    for (i = 0; i < RTL_NUMBER_OF(LogDevice->Io); i++)
+    {
+        LogDevice->Io[i].CurrentBase = 0;
+    }
+    for (i = 0; i < RTL_NUMBER_OF(LogDevice->Irq); i++)
+    {
+        LogDevice->Irq[i].CurrentNo =
+        LogDevice->Irq[i].CurrentType = 0;
+    }
+    for (i = 0; i < RTL_NUMBER_OF(LogDevice->Dma); i++)
+    {
+        LogDevice->Dma[i].CurrentChannel = 4;
+    }
+    for (i = 0; i < RTL_NUMBER_OF(LogDevice->MemRange); i++)
+    {
+        LogDevice->MemRange[i].CurrentBase =
+        LogDevice->MemRange[i].CurrentLength = 0;
+    }
+    for (i = 0; i < RTL_NUMBER_OF(LogDevice->MemRange32); i++)
+    {
+        LogDevice->MemRange32[i].CurrentBase =
+        LogDevice->MemRange32[i].CurrentLength = 0;
+    }
+}
+
+static
+CODE_SEG("PAGE")
+VOID
+WriteResources(
+    _In_ PUCHAR ReadDataPort,
+    _In_ PISAPNP_LOGICAL_DEVICE LogDevice,
+    _In_ PCM_PARTIAL_RESOURCE_LIST PartialResourceList)
+{
+    UCHAR i,
+          NumberOfIo = 0,
+          NumberOfIrq = 0,
+          NumberOfDma = 0,
+          NumberOfMemory = 0,
+          NumberOfMemory32 = 0;
+
+    PAGED_CODE();
+
+    WriteLogicalDeviceNumber(LogDevice->LDN);
+
+    for (i = 0; i < PartialResourceList->Count; i++)
+    {
+        PCM_PARTIAL_RESOURCE_DESCRIPTOR Descriptor = &PartialResourceList->PartialDescriptors[i];
+        UCHAR Index;
+
+        switch (Descriptor->Type)
+        {
+            case CmResourceTypePort:
+            {
+                (VOID)FindIoDescriptor(LogDevice,
```

Comment:
Suppose finding the descriptor fails; what then should happen?

Reply:
The second call never fails.